Style disabled textinput and textarea

oorestisime opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
Styles disabled TextInput TextArea. as a side-effect Select is also styled when disabled
#### Where should the reviewer start?

I wonder whether we should have opacity style as a helper function and whether opacity style should first live in global section of theme. Like all the disabled things should be styled with the same way and then if user decides to override then ok.
it should remove some code-duplication from this PR. Didn't do this right now since currently we don't handle things in this way

#### What testing has been done on this PR?
Jest and manual on storybook
#### How should this be manually tested?
storybook
#### Any background context you want to provide?

#### What are the relevant issues?
#2825 
#### Screenshots (if appropriate)

#### Do the grommet docs need to be updated?
done
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
compatible

ShimiSun commented on src/js/themes/base.js:
```diff
@@ -783,12 +783,18 @@ export const generate = (baseSpacing = 24, scale = 6) => {
-    //   extend: undefined,
-    // },
+    textArea: {
+      disabled: {
```
I'm thinking maybe we want the `extend` object to be available in `disabled`. hence the user can customize it as he wishes. thoughts?

ShimiSun replied:
Do you think it will be tricky?

ShimiSun commented on src/js/themes/base.js:
```diff
@@ -783,12 +783,18 @@ export const generate = (baseSpacing = 24, scale = 6) => {
+    },
+    textInput: {
+      disabled: {
+        opacity: 0.3,
```
where did the `0.3` come from? is it because we use this value for button disable? (which will make sense)

oorestisime replied:
yes it was from the button.

ShimiSun commented:
> I wonder whether we should have opacity style as a helper function and whether opacity style should first live in global section of theme. Like all the disabled things should be styled with the same way and then if user decides to override then ok.
I agree. it feels wrong that we have `opacity: 0.3` scattered around the theme rather than being placed in a dedicated disabled logic.

oorestisime commented:
So i ve reworked this PR.

- added disabledStyle in global
- removed both form textinput, textarea and button (but i kept backwards compat)
- updated docs

ShimiSun commented on src/js/components/Button/doc.js:
```diff
@@ -134,7 +134,8 @@ export const themeDoc = {
     type: 'string | { dark: string, light: string }',
   },
   'button.disabled.opacity': {
```
I'm thinking we shouldn't expose it at all, but still support it in the background. thoughts?

oorestisime replied:
well if we still support this i think it is better to write it down. enables people to see the deprecation wanring as well. but no strong feelings here.

ericsoderberghp replied:
Are we sure we want to deprecate this? I do like the idea of having something more general, but I could still see a per-component style for this to customize further.

oorestisime replied:
i was thinking to deprecate this mainly to match the rest of the areas the disabled opacity will be used. I mean why would we have a button specific one but not one for textInput or textArea? Unless we do add to the other areas as well.

ericsoderberghp replied:
I see. I think I'd tend towards each component getting their own, rather than removing it from Button.

ShimiSun commented on src/js/themes/base.js:
```diff
@@ -785,12 +785,18 @@ export const generate = (baseSpacing = 24, scale = 6) => {
+      disabled: {
+        opacity: 0.3,
+      },
+      extend: undefined,
```
Can we comment this line out? so we'll be consistent with the other objects

ShimiSun commented on src/js/components/TextArea/StyledTextArea.js:
```diff
@@ -1,6 +1,11 @@
+  focusStyle,
+  inputStyle,
+  placeholderStyle,
+  disabledStyle,
```
can we please move `disabledStyle,` to line 4 and maintain the right order.

oorestisime replied:
yeap. also did for input and button

ShimiSun commented on src/js/themes/base.js:
```diff
@@ -785,12 +785,18 @@ export const generate = (baseSpacing = 24, scale = 6) => {
+      disabled: {
+        opacity: 0.3,
+      },
+      extend: undefined,
```
Can we comment this line out? so we'll be consistent with the other objects

ShimiSun commented on src/js/utils/styles.js:
```diff
@@ -252,3 +252,8 @@ export const genericStyles = css`
     )}
 `;
+
+export const disabledStyle = componentStyle => css`
```
This is so much better 💯

ShimiSun commented:
`disabled` prop is missing from corresponding index.d.ts files.

oorestisime commented:
disabled is part of the intrinsic elements of input. i am not sure we need to add it

ericsoderberghp commented on src/js/themes/base.js:
```diff
@@ -212,6 +212,9 @@ export const generate = (baseSpacing = 24, scale = 6) => {
           color: 'focus',
         },
       },
+      disabled: {
```
I think it might be more consistent to put this under `global.control.disabled`, since that is where the border for TextInput, TextArea, and Select are all defined. And, `disabled` feels more like it should be associated with a control context than with everything.

ericsoderberghp commented on src/js/themes/base.js:
```diff
@@ -299,6 +299,7 @@ export const generate = (baseSpacing = 24, scale = 6) => {
       primary: {
         // color: { dark: undefined, light: undefined }
       },
+      disabled: undefined,
```
I think this should be commented out, like the other undefined values in this file.

oorestisime replied:
done!
